Report CPU info to HTCondor monitoring

bbockelm opened the pull request:
This PR fixes #15763 and #15764 by adding CPU information to the HTCondor monitoring.  In particular, it adds:
- CPU time usage.
- CPU model information.
- CPU time usage from child processes.

bbockelm committed:
Report CPU usage information via HTCondor service.

This adds CPU usage information to the HTCondor service reporting;
it reports the entire CPU (user + system) usage for this process.
As the "job CPU" information reported by the framework excludes the
framework startup costs, the "total CPU" is slightly more than what
the Timing service prints on stdout.

bbockelm committed:
Report CPU model to HTCondor.

This commit has the framework forward the CPU model information to the
HTCondor monitoring.

bbockelm committed:
Include child processes in CPU timing.

With this commit, getCPU() additionally returns all usage information
from child processes.

cmsbuild commented:
A new Pull Request was created by  @bbockelm (Brian Bockelman) for CMSSW_8_1_X.

It involves the following packages:

FWCore/Services
FWCore/Utilities


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @makortel, @wddgit, @wmtan this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Services/plugins/CPU.cc:
```diff
@@ -30,18 +31,21 @@
       int totalNumberCPUs_;
       double averageCoreSpeed_;
       bool reportCPUProperties_;
-      
```
Updating the member data in the cpuInfo call is not thread safe. As far as I can tell, there is no need for `totalNumberCPUs_` or `averageCoreSpeed_` to be member data. It also looks like `reportCPUProperties_` could be `const`.

bbockelm replied:
Done.

Dr15Jones commented on FWCore/Services/plugins/CPU.cc:
```diff
@@ -125,9 +129,20 @@ namespace edm {
+        return cpuInfoImpl(models, avgSpeed, nullptr);
+    }
+
+    bool CPU::cpuInfoImpl(std::string &result_models, double &result_avg_speed, Service<JobReport>* reportSvc)
```
This seems to do a lot more than just the work need to return `result_models` and `result_avg_speed`. I think the function should be better factored so that it doesn't know about the `JobReport`.

bbockelm replied:
Done.

cmsbuild commented:
+1
Tested at: f3c060515c74aa620213915a6beead6a8a9152ec
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15767/15144/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15767/15144/summary.html

bbockelm committed:
Cleanly separate cpuinfo parsing from statistics collection.

Now, one function will parse the cpuinfo file - each different
statistic we want will utilize this information in a separate
method.

cmsbuild commented:
Pull request #15767 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

bbockelm committed:
Merge remote-tracking branch 'official-cmssw/CMSSW_8_1_X' into report_cpu_info

Conflicts:
	FWCore/Services/plugins/CPU.cc

cmsbuild commented:
Pull request #15767 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e8d9da8fd751392c84b556d8861e07abe8ef02de
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15767/15579/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15767/15579/summary.html

Dr15Jones commented:
@bbockelm how do we know if this is working?

bbockelm commented:
Hi Chris,

Add the following two lines to your pset:
```
process.CPU = cms.Service("CPU", reportCPUProperties=cms.untracked.bool(True))
process.CondorStatusService = cms.Service("CondorStatusService", tag=cms.untracked.string("DIGI"), debug=cms.untracked.bool(True))
```

You should see the condor invocations echoed to stdout and the CPU information in the FJR.

Brian

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

davidlange6 commented:
+1

dan131riley commented:
Hi @bbockelm 

If I add

```
process.Timing = cms.Service("Timing")
process.CPU = cms.Service("CPU", reportCPUProperties=cms.untracked.bool(True))
process.CondorStatusService = cms.Service("CondorStatusService", tag=cms.untracked.string("DIGI"), debug=cms.untracked.bool(True))
```

to a job in CMSSW_8_1_0_pre13 (that has this PR) I get a failure

```
----- Begin Fatal Exception 24-Oct-2016 12:55:33 EDT-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing service of type CondorStatusService
Exception Message:
MissingParameter: The required parameter 'summaryOnly' was not specified.
----- End Fatal Exception -------------------------------------------------
```

which looks like the TimingService is getting constructed before fillDescriptions has been called...this does not happen if the TimingService isn't included in the config. Found by a CRAB user, see this HN and the upstream thread:

https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/2205/2/1.html

Since this happens while constructing CondorStatusService, I strongly suspect this PR is responsible.

dan131riley commented:
Never mind, I had missed that this was fixed by #16311
